### PR TITLE
[Darwin] Further restrict inference of the simulator environment

### PR DESCRIPTION
**Pull request by DougGregor**
The *-simulator target triples have been used mostly consistently in build systems for
several years to indicate simulator targets... except for Swift's own CMake configuration
and `build-script`, which was still not doing it until recently. There are still `.swiftinterface`
files around that depend on the inference.

Rather than outright removing inference of the "simulator" environment, move it into
the driver and warn that it's happening, so other tools can move off it over time. This
ensures that all frontend/linker/etc. jobs see the "simulator" environment, so the hack is
only in a single place.

This is as far as we can go toward completely removing the hack (rdar://problem/35810403),
because of the need to continue supporting existing `.swiftinterface` files.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by harlanhaskins** [comment]
Seems like we should also remove this from `swift-driver`. https://github.com/apple/swift-driver/blob/master/Sources/SwiftDriver/Utilities/Triple%2BPlatforms.swift#L133

**Comment by DougGregor** [comment]
Linux failure is unrelated. Trying again...

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please smoke test Linux

**Comment by DougGregor** [comment]
GitHub connectivity problems affected the compatibility suite. Sigh.

**Comment by DougGregor** [comment]
Sigh. I need to split this pull request into "fix our build system / test" and "remove the cruft".

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please clean test

**Comment by DougGregor** [comment]
@swift-ci please clean test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please clean test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
Much of this pull request came in via https://github.com/apple/swift/pull/30843. I'll rebase the rest on top of this.

File: cmake/modules/SwiftConfigureSDK.cmake [hunk 1]
```diff
@@ -172,6 +174,13 @@ macro(configure_sdk_darwin
     "${SWIFT_SDK_${prefix}_ARCHITECTURES}"        # rhs
     SWIFT_SDK_${prefix}_MODULE_ARCHITECTURES)     # result
 
+  # Determine whether this is a simulator SDK.
+  if ( ${xcrun_name} MATCHES "simulator" )
```
Comment (compnerd) [inline cmake/modules/SwiftConfigureSDK.cmake:178]: The convention in LLVM (and Swift) is not have spaces after the `if` and not to have spaces inside the `(`.
```suggestion
  if(xcrun_name MATCHES simulator)
```

File: cmake/modules/SwiftConfigureSDK.cmake [hunk 2]
```diff
@@ -181,6 +190,12 @@ macro(configure_sdk_darwin
     set(SWIFT_SDK_${prefix}_ARCH_${arch}_TRIPLE
         "${arch}-apple-${SWIFT_SDK_${prefix}_TRIPLE_NAME}")
 
+    # If this is a simulator target, append -simulator.
+    if (SWIFT_SDK_${prefix}_IS_SIMULATOR)
```
Comment (compnerd) [inline cmake/modules/SwiftConfigureSDK.cmake:194]:
```suggestion
    if(SWIFT_SDK_${prefix}_IS_SIMULATOR)
```

File: cmake/modules/SwiftConfigureSDK.cmake [hunk 3]
```diff
@@ -181,6 +190,12 @@ macro(configure_sdk_darwin
...
+      set(SWIFT_SDK_${prefix}_ARCH_${arch}_TRIPLE
+          "${SWIFT_SDK_${prefix}_ARCH_${arch}_TRIPLE}-simulator")
+    endif ()
```
Comment (compnerd) [inline cmake/modules/SwiftConfigureSDK.cmake:197]:
```suggestion
    endif()
```

File: cmake/modules/SwiftConfigureSDK.cmake (hunk 1 above)
Comment (DougGregor) [reply cmake/modules/SwiftConfigureSDK.cmake:178]: I had no idea, thank you!

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci test platform Linux

**Comment by DougGregor** [comment]
@swift-ci smoke test Linux

**Comment by DougGregor** [comment]
@swift-ci smoke test Linux

**Comment by fredriss** [comment]
When this was merged, the Linux PR testing started failing in lldb tests. First Failure here: https://ci.swift.org/job/oss-swift-package-linux-ubuntu-16_04/4933/

I have no idea what's going on, it doesn't make much sense, but the revert unblocked things.

**Pull request merged**